### PR TITLE
cleanup validations for VPN connection creation

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

This PR adds a validation when creating a VPN connection in a VPC to make sure the VPN gateway exists.

Fixes: #7250 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [x] Trivial

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Review by rohityadavcloud — APPROVED** [review]
LGTM, this may need s2s testing if not already covered by a smoketest

**Review by hsato03 — APPROVED** [review]
CLGTM.

Just one remark, maybe we can use the same error message in the `getAndValidate` methods.

**Comment by DaanHoogland** [comment]
> CLGTM.
> 
> Just one remark, maybe we can use a unique error message in the `getAndValidate` methods.

The calls have different messages passed (though only slightly). Can you specify what you would like to see there @hsato03 ?

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by hsato03** [comment]
> > CLGTM.
> > Just one remark, maybe we can use a unique error message in the `getAndValidate` methods.
> 
> The calls have different messages passed (though only slightly). Can you specify what you would like to see there @hsato03 ?

What about `throw new InvalidParameterValueException(String.format("Unable to find the specified Site-to-Site VPN customer gateway: %s.", customerGatewayId));` in the `getAndValidateSite2SiteCustomerGateway`

and `throw new InvalidParameterValueException(String.format("Unable to find the specified Site-to-Site VPN gateway %s.", vpnGatewayId));` in the `getAndValidateSite2SiteVpnGateway`?

This should cover all cases.

**Comment by DaanHoogland** [comment]
@hsato03 applied your suggestion and did some extra cleanup.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test alma9 kvm-alma9

File: server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java [hunk 1]
```diff
@@ -326,13 +308,53 @@ public Site2SiteVpnConnection createVpnConnection(CreateVpnConnectionCmd cmd) th
         return conn;
     }
 
+    @NotNull
+    private Site2SiteCustomerGateway getAndValidateSite2SiteCustomerGateway(Long customerGatewayId, Account caller) {
+        Site2SiteCustomerGateway customerGateway = _customerGatewayDao.findById(customerGatewayId);
+        if (customerGateway == null) {
+            throw new InvalidParameterValueException(String.format("Unable to found specified Site to Site VPN customer gateway %s !", customerGatewayId));
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java:315]:
```suggestion
            throw new InvalidParameterValueException(String.format("Unable to find specified Site to Site VPN customer gateway %s !", customerGatewayId));
```

File: server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java [hunk 2]
```diff
@@ -326,13 +308,53 @@ public Site2SiteVpnConnection createVpnConnection(CreateVpnConnectionCmd cmd) th
...
+        }
+        _accountMgr.checkAccess(caller, null, false, customerGateway);
+        return customerGateway;
+    }
+
+    @NotNull
+    private Site2SiteVpnGateway getAndValidateSite2SiteVpnGateway(Long vpnGatewayId, Account caller) {
+        Site2SiteVpnGateway vpnGateway = _vpnGatewayDao.findById(vpnGatewayId);
+        if (vpnGateway == null) {
+            throw new InvalidParameterValueException(String.format("Unable to found specified Site to Site VPN gateway %s !", vpnGatewayId));
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java:325]:
```suggestion
            throw new InvalidParameterValueException(String.format("Unable to find specified Site to Site VPN gateway %s !", vpnGatewayId));
```

**Review by sureshanaparti — APPROVED** [review]
minor suggestions, clgtm

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java [hunk 3]
```diff
@@ -326,13 +314,53 @@ public Site2SiteVpnConnection createVpnConnection(CreateVpnConnectionCmd cmd) th
...
+    }
+
     @Override
     @DB
     @ActionEvent(eventType = EventTypes.EVENT_S2S_VPN_CONNECTION_CREATE, eventDescription = "starting s2s vpn connection", async = true)
     public Site2SiteVpnConnection startVpnConnection(long id) throws ResourceUnavailableException {
         Site2SiteVpnConnectionVO conn = _vpnConnectionDao.acquireInLockTable(id);
         if (conn == null) {
-            throw new CloudRuntimeException("Unable to acquire lock on " + conn);
+            throw new CloudRuntimeException("Unable to acquire lock for starting on connection with id " + id);
```
Comment (vishesh92) [inline server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java:363]:
```suggestion
            throw new CloudRuntimeException("Unable to acquire lock for starting vpn connection with id " + id);
```

File: server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java [hunk 4]
```diff
@@ -578,7 +597,7 @@ public boolean deleteVpnConnection(DeleteVpnConnectionCmd cmd) throws ResourceUn
     private void stopVpnConnection(Long id) throws ResourceUnavailableException {
         Site2SiteVpnConnectionVO conn = _vpnConnectionDao.acquireInLockTable(id);
         if (conn == null) {
-            throw new CloudRuntimeException("Unable to acquire lock on " + conn);
+            throw new CloudRuntimeException("Unable to acquire lock for stopping on connection with id " + id);
```
Comment (vishesh92) [inline server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java:600]:
```suggestion
            throw new CloudRuntimeException("Unable to acquire lock for stopping vpn connection with id " + id);
```

File: server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java (hunk 3 above)
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java:363]:
```suggestion
            throw new CloudRuntimeException("Unable to acquire lock for starting of VPN connection with ID " + id);
```

File: server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java (hunk 4 above)
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java:600]:
```suggestion
            throw new CloudRuntimeException("Unable to acquire lock for stopping of VPN connection with id " + id);
```

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan LLtest alma9 kvm-alma9

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Review by shwstppr — APPROVED** [review]
Cdde LGTM

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Review by rajujith — CHANGES_REQUESTED** [review]
The issue is not resolved.
When I try creating a site to site VPN connection without enabling the VPN gateway of the VPC it doesn't throw any error.

**Comment by sureshanaparti** [comment]
> The issue is not resolved. When I try creating a site to site VPN connection without enabling the VPN gateway of the VPC it doesn't throw any error.

thanks for testing @rajujith

**Comment by DaanHoogland** [comment]
> The issue is not resolved. When I try creating a site to site VPN connection without enabling the VPN gateway of the VPC it doesn't throw any error.

hm, it has been a while and had some updates. I'll retest and see...

File: server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java [hunk 5]
```diff
@@ -404,7 +425,7 @@ protected boolean doDeleteCustomerGateway(Site2SiteCustomerGateway gw) {
 
     protected void doDeleteVpnGateway(Site2SiteVpnGateway gw) {
         List<Site2SiteVpnConnectionVO> conns = _vpnConnectionDao.listByVpnGatewayId(gw.getId());
-        if (conns != null && conns.size() != 0) {
+        if (CollectionUtils.isEmpty(conns)) {
```
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java:428]:
```suggestion
        if (!CollectionUtils.isEmpty(conns)) {
```

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Review by GaOrtiga — APPROVED** [review]
CLGTM

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

**Comment by DaanHoogland** [comment]
> > The issue is not resolved. When I try creating a site to site VPN connection without enabling the VPN gateway of the VPC it doesn't throw any error.
> 
> hm, it has been a while and had some updates. I'll retest and see...

@rajujith , this was probably due to some merge issues, It should work now. I.E. you should not be able to create a VPN connection without having a VPN gateway available. Can you try?

**Comment by kiranchavala** [comment]
> > > The issue is not resolved. When I try creating a site to site VPN connection without enabling the VPN gateway of the VPC it doesn't throw any error.
> > 
> > 
> > hm, it has been a while and had some updates. I'll retest and see...
> 
> @rajujith , this was probably due to some merge issues, It should work now. I.E. you should not be able to create a VPN connection without having a VPN gateway available. Can you try?

@DaanHoogland 

I tried again with the packages from 17th Jul and the issue is not resolved.

When I try creating a site to site VPN connection without enabling the VPN gateway of the VPC it doesn't throw any error.

**Comment by DaanHoogland** [comment]
> > > > The issue is not resolved. When I try creating a site to site VPN connection without enabling the VPN gateway of the VPC it doesn't throw any error.
> > > 
> > > 
> > > hm, it has been a while and had some updates. I'll retest and see...
> > 
> > 
> > @rajujith , this was probably due to some merge issues, It should work now. I.E. you should not be able to create a VPN connection without having a VPN gateway available. Can you try?
> 
> @DaanHoogland
> 
> I tried again with the packages from 17th Jul and the issue is not resolved.
> 
> When I try creating a site to site VPN connection without enabling the VPN gateway of the VPC it doesn't throw any error.

@kiranchavala it was only caught/solved in the API. It is now also the UI.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Review by rajujith — APPROVED** [review]
LGTM. It handles the error condition now.

File: ui/src/views/network/VpcTab.vue [hunk 6]
```diff
@@ -795,12 +795,12 @@ export default {
 
       this.formRef.value.validate().then(() => {
         const values = toRaw(this.form)
-
-        api('createVpnConnection', {
-          s2svpngatewayid: this.vpnGateways[0].id,
+        const params = {
+          s2svpngatewayid: this.vpnGateways[0] ? this.vpnGateways[0].id : null,
```
Comment (vishesh92) [inline ui/src/views/network/VpcTab.vue:799]: `s2svpngatewayid` is a required parameter for `createVpnConnection` command.

**Review by vishesh92 — APPROVED** [review]
clgtm

**Pull request merged**